executor/aggregate.go

```diff
@@ -862,6 +872,50 @@ func (e *HashAggExec) getPartialResults(groupKey string) []aggfuncs.PartialResul
 	return partialResults
 }
 
+func (e *HashAggExec) executeFirstRow(ctx context.Context, chk *chunk.Chunk) (err error) {
+	for {
+		mSize := e.childResult.MemoryUsage()
+		if e.childResult == nil || e.firstRowProcessed == e.childResult.NumRows() {
+			err := Next(ctx, e.children[0], e.childResult)
+			e.memTracker.Consume(e.childResult.MemoryUsage() - mSize)
+			if err != nil {
+				return err
+			}
+			e.firstRowProcessed = 0
+		}
+
+		failpoint.Inject("unparallelHashAggError", func(val failpoint.Value) {
+			if val.(bool) {
+				failpoint.Return(errors.New("HashAggExec.unparallelExec error"))
+			}
+		})
+
+		// no more data.
+		if e.childResult.NumRows() == 0 {
+			return nil
+		}
```

Comment:
The check can move to the block `if e.childResult == nil || e.firstRowProcessed == e.childResult.NumRows() {...}`